drafts/bitcartel-payment-disclosure/draft.rst

```diff
@@ -0,0 +1,336 @@
+
+Also note that an independent third party cannot know for sure if the sender and recipient are not colluding to hide value transfer.  JoinSplits have two inputs and two outputs.  The recipient may identify their receiving address for one output, but not disclose the fact that the other output also belongs to them.
+
+TODO: What happens if a transaction is created (at which point a payment disclosure can be created) but never gets mined and is stuck in the mempool?  Should the payment disclosure be retained or purged?
+
+.. [KDFT] https://github.com/zcash/zcash/issues/2102
+
+
+Specification
+-------------
+When creating a shielded transaction, for each JoinSplit output, a data structure is created to record the following fields:
+
+- transaction id
+- index [0..len-1] of JoinSplit in array of JoinSplits contained in transaction
+- index [0..1] of JoinSplit output
+- recipient's payment address is a shielded address ``(a_pk, pk_enc)`` [PROTOCOL] §3.1 Payment Addresses and Keys
+- symmetric key used to encrypt the note, also referred to as the ephemeral secret key ``esk`` [PROTOCOL] §4.10.1 Generate a new KA (public, private) key pair ``(epk, esk)``.
+- JoinSplitSig private key used to sign the JoinSplit transaction ``joinSplitPrivKey`` [PROTOCOL] §4.4 Sending Notes
+
+The payment disclosure data should be persisted to disk or a database so it can be retrieved later.
+
+When persisting, third party applications should expect serialization of payment disclosure to follow Zcash and upstream convention.
+
+::
+
+    ADD_SERIALIZE_METHODS;
+    ...
+    READWRITE(txid);
+    ...
+    READWRITE(joinSplitPrivKey);
+
+A new RPC call will be introduced to allow the sender to retrieve the payment disclosure data for a given shielded output index.
+
+``z_getpaymentdisclosure txid joinsplit_index output_index [message]``
+
+- Returns the payment disclosure in hexadecimal format.
+- Message is an optional parameter, a UTF-8 string.  We may want to restrict/sanitize this user input, e.g. number of characters, allowed characters.
+
+The sender wants the payment disclosure to be non-malleable, to prevent an attacker modifying details like the refund address.  To achieve this, the sender will sign the payment disclosure with the JoinSplitSig private key and append the signature to the end of the payment disclosure data.
+
+A new RPC call will be introduced to allow a third party to verify a payment disclosure.
+
+``z_validatepaymentdisclosure paymentdisclosure``
+
+Validates a payment disclosure and returns JSON output as follows:
+
+::
+
+    {
+    valid : true,
+    rawhex : "0011...ff",
+    version: 0
+    txid: "4e...",
+    jsindex: 0,
+    outputindex: 1,
+    paymentaddress: zABC123...,
+    value: 10.56100002
+    message: "pay me at z007",
+    signature: "10ab23..."
+    }
+
+Valid field is true if all the following conditions hold:
```

Comment:
No checking of the note commitment, as in  [PROTOCOL] §4.10.2?

Reply:
Now fixed.